bulk_deploy/tsuru_deploy_client.rb

```diff
@@ -0,0 +1,105 @@
+require 'fileutils'
+
+class TsuruDeployClient
+
+  attr_accessor :api_client, :logger
+
+  def initialize(
+    api_client:, logger:, environment:, host:, tsuru_home:, protocol: "https://"
+  )
+    @api_client = api_client
+    @logger = logger
+    @tsuru_home = tsuru_home
+
+    target = URI.parse(protocol + environment + "-api." + host)
+    target_list = `tsuru target-list`
+
+    unless target_list.include?(target.to_s)
+      if !system("tsuru target-add #{environment} #{target}")
+        raise "Failed to add the target"
+      end
+    end
+
+    unless target_list.include?("* #{environment} #{target.to_s}")
+      if !system("tsuru target-set #{environment}")
+        raise "Failed to set the target"
+      end
+    end
+  end
+
+  def deploy_app(user:, app:, env_vars: {}, postgres: '', git: false, units: 3)
+    self.logger.info("Going to deploy #{app[:name]}")
+    self.logger.info("Login user #{user[:email]} of the team #{user[:team]}")
+    self.api_client.login(user[:email], user[:password])
+
+    if not self.api_client.list_apps().include? app[:name]
+      self.logger.info("Create application #{app[:name]} " \
+        "on the platform #{app[:platform]}")
+      self.api_client.create_app(app[:name], app[:platform])
+    end
+
+    # Set environment variables, if needed
+    if env_vars.length > 0
+      env_vars.each do |key,value|
+        self.api_client.set_env_var(app[:name], key, value)
+      end
+    end
+
+    if postgres != ''
+      instance_name = postgres
+      unless self.api_client.list_service_instances().include? instance_name
+          self.logger.info("Add postgres service instance #{instance_name}")
+          self.api_client.add_service_instance("postgresql", instance_name)
+      end
+
+      unless self.api_client.app_has_service(app[:name], instance_name)
+        self.logger.info("Bind service #{instance_name} to #{app[:name]}")
+        self.api_client.bind_service_to_app(instance_name, app[:name])
+      end
+    end
+
+    if git
+      self.logger.info("Deploy #{app[:name]} via git")
+      git_deploy(
+        app[:dir],
+        self.api_client.get_app_repository(app[:name]),
+        user[:key]
+      )
+    else
+      self.logger.info("Deploy #{app[:name]} via app-deploy")
+      app_deploy(app[:dir], app[:name])
+    end
+
+    deployed_units = self.api_client.get_app_info(app[:name])["units"].length
+    if deployed_units < units
+      self.api_client.add_units(units - deployed_units, app[:name])
```

Comment:
I hit a race condition here once, caused because issues in tsuru itself:

```
remote: ---- Adding routes to 1 new units ----
remote:  ---> Added route to unit 306bc64b3f
remote: 
remote: OK
To git@hector200715-gandalf.tsuru2.paas.alphagov.co.uk:flask-app-0982757341.git
 * [new branch]      master -> master
Identity removed: /tmp/tsuru_tmp/.ssh/id_rsa_testuser8@testteam9651757341.co.uk ( keymon@keyair.local)
E, [2015-07-22T14:50:06.568101 #12013] ERROR -- : Request: PUT /apps/flask-app-0982757341/units 2
E, [2015-07-22T14:50:06.568216 #12013] ERROR -- : Response Body: App locked by testuser8@testteam9651757341.co.uk, running PUT /apps/flask-app-0982757341/units. Acquired in 2015-07-22T13:48:56Z

I, [2015-07-22T14:50:06.582189 #12013]  INFO -- : Write state file /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/deploy.state
/Users/keymon/gds/tsuru-performance-tests/bulk_deploy/tsuru_api_client.rb:423:in `throw': uncaught throw #<Net::HTTPConflict 409 Conflict readbody=true> (UncaughtThrowError)
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/tsuru_api_client.rb:423:in `request'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/tsuru_api_client.rb:353:in `request_json'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/tsuru_api_client.rb:201:in `add_units'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/tsuru_deploy_client.rb:75:in `deploy_app'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/deploy_actions.rb:130:in `block in apply'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/deploy_actions.rb:102:in `each'
	from /Users/keymon/gds/tsuru-performance-tests/bulk_deploy/deploy_actions.rb:102:in `apply'
	from deploy.rb:84:in `<main>'
```

But I am not sure what was locking it :-m

It seems the deployment scripts (app deploy or git push) can return